src/FilenameParsing/FileIDHelperResolutionStrategy.php

```diff
@@ -363,52 +389,78 @@ public function buildFileID($tuple)
-        array_unshift($helpers, $this->getDefaultFileIDHelper());
+        $defaultHelper = $this->getDefaultFileIDHelper();
+        if (!in_array($defaultHelper, $helpers)) {
+            array_unshift($helpers, $defaultHelper);
+        }
 
         /** @var FileIDHelper[] $resolvableHelpers */
         $resolvableHelpers = [];
 
         // Search for a helper that will allow us to find a file
         foreach ($helpers as $helper) {
             try {
-                $fileID = $helper->buildFileID(
-                    $parsedFileID->getFilename(),
-                    $parsedFileID->getHash()
-                );
+                $fileID = $helper->buildFileID($parsedFileID->getFilename(), $parsedFileID->getHash());
 
                 if ($filesystem->has($fileID) && $this->validateHash($helper, $parsedFileID, $filesystem)) {
                     $resolvableHelpers[] = $helper;
                 }
             } catch (InvalidArgumentException $ex) {
-                // Our helper couldn't build a FileID with the provided arguments, that means it's not valid helper for
+                // Our helper couldn't build a FileID with the provided arguments, that means it's not a valid helper for
                 // this tuple.
             }
         }
 
         foreach ($resolvableHelpers as $helper) {
-            // Make sure our yield file has an hash
+            // Make sure our yield file has a hash
             $hash = $parsedFileID->getHash() ?: $this->findHashOf($helper, $parsedFileID, $filesystem);
 
             $folder = $helper->lookForVariantIn($parsedFileID);
             $possibleVariants = $filesystem->listContents($folder, true);
+            $possibleVariants = array_filter($possibleVariants, function ($possibleVariant) {
+                return $possibleVariant['type'] !== 'dir';
+            });
+            $possibleVariants = array_map(function ($possibleVariant) {
+                return $possibleVariant['path'];
+            }, $possibleVariants);
+
+            $mainVariant = $helper->buildFileID($parsedFileID->getFilename(), $parsedFileID->getHash());
+            if (!in_array($mainVariant, $possibleVariants) && $filesystem->has($mainVariant)) {
+                $possibleVariants[] = $mainVariant;
+            }
+
             foreach ($possibleVariants as $possibleVariant) {
-                if ($possibleVariant['type'] !== 'dir' && $helper->isVariantOf($possibleVariant['path'], $parsedFileID)) {
-                    yield $helper->parseFileID($possibleVariant['path'])->setHash($hash);
+                if ($helper->isVariantOf($possibleVariant, $parsedFileID)) {
+                    yield $helper->parseFileID($possibleVariant)->setHash($hash);
                 }
             }
         }
     }
-    
+
     public function cleanFilename($filename)
     {
         return $this->getDefaultFileIDHelper()->cleanFilename($filename);
     }
-    
+
+    /**
+     * todo: needs documenting
```

Comment:
Document

Reply:
Those methods that didn't have a PHPDoc block are implemented from an interface that has the doc block. The interface doc block are inherited.